evaluation/what-to-support.md

```diff
@@ -0,0 +1,152 @@
+# What should we support
+
+There are a lot of things we would ideally support, this file is intended as a collection of all of them. Not all of them will end up being supported as they may conflict with other design and implementation goals.
+
+## purely const
+
+Dealing with `const`-ness does not change type signatures.
+
+```rust
+impl<const N: usize, T> [T; N] {
+    // `map` is const if `F` is const.
+    //
+    // This is the case as `map` eagerly uses `f` in
+    // its function body.
+    const<C> fn map<U, F, const<C>>(self, f: F) -> [U; N]
```

Comment:
```suggestion
    const<C> fn map<U, F>(self, f: F) -> [U; N]
```

So far we've been using the syntax without repeating the `C` in the parameter list. For a while we put the generic declaration in a `for`-binding, but we dropped that as the feedback from the lang team was that the syntax was too noisy. I assume the second `const<C>` here exists so that the generic is declared?

Reply:
To illustrate, we used to have this:
```rust
for<effect C> const<C> fn map<U, F>(self, f: F) -> [U; N]
```
The generic is defined _before_ usage in the keyword. But this syntax was considered to be syntactically overwhelming. Instead we opted to just make the `for`-binding implied, and instead take a "using it == declaring it" approach.

Reply:
yeah :thinking: also don't think that this is the syntax we should use, but having a maximally explicit syntax makes it imo easier to think/talk about stuff.

With implicit effect parameters everyone has to have the same assumptions/understanding for how they get actually desugared.